pilot/pkg/serviceregistry/kube/controller/endpoint_builder.go

```diff
@@ -73,6 +93,38 @@ func (b *EndpointBuilder) buildIstioEndpoint(
 		Address:         endpointAddress,
 		EndpointPort:    uint32(endpointPort),
 		ServicePortName: svcPortName,
-		Network:         b.controller.endpointNetwork(endpointAddress),
+		Network:         b.endpointNetwork(endpointAddress),
 	}
 }
+
+// return the mesh network for the endpoint IP. Empty string if not found.
+func (b *EndpointBuilder) endpointNetwork(endpointIP string) string {
+	// Try to determine the network by checking whether the endpoint IP belongs
+	// to any of the configure networks' CIDR ranges
+	if b.controller.ranger != nil {
+		entries, err := b.controller.ranger.ContainingNetworks(net.ParseIP(endpointIP))
+		if err != nil {
+			log.Errora(err)
+			return ""
+		}
+		if len(entries) > 1 {
+			log.Warnf("Found multiple networks CIDRs matching the endpoint IP: %s. Using the first match.", endpointIP)
+		}
+		if len(entries) > 0 {
+			return (entries[0].(namedRangerEntry)).name
+		}
+	}
+
+	// If not using cidr-lookup, or non of the given ranges contain the address, use the pod-label
+	if nw := b.labels[label.IstioNetwork]; nw != "" {
+		return nw
+	}
+
+	// If we're building the endpoint based on proxy meta, prefer the injected ISTIO_META_NETWORK value.
+	if b.metaNetwork != "" {
```

Comment:
My concern is the meta network can conflict with the above labels, which is confusing, is there any docs on this?